### ISSUE TITLE
detect_vm() is run before xenfs is mounted, leading to wrong Xen dom0 detection

### ISSUE THREAD

**Issue by vliaskov** [systemd/systemd#6442]
### Submission type

 - Bug report

### systemd version the issue has been seen with

systemd-228, but looks similar in systemd upstream

### Used distribution

SLES12SP2

### In case of bug report: Expected behaviour you didn't see

ConditionVirtualisation=none on Xen dom0 (as expected with fix #2625)

### In case of bug report: Unexpected behaviour you saw

ConditionVirtualisation=xen is detected and cached by systemd[1] on Xen dom0 bootup.

This is an example log where systemd[1] detects and caches the wrong value (cached_found= VIRTUALIZATION_XEN), presumably because xen procfs has not been mounted yet. systemd-hostnamedd[2202] which runs later detects the correct type (cached_found=VIRTUALIZATION_NONE) for Xen dom0:

Jul 13 16:47:09 linux-usly kernel: Booting paravirtualized kernel on Xen
16:47:14 systemd[1]: Virtualization XEN not found, /proc/xen/capabilities does not exist
16:47:14 systemd[1]: Virtualization xen found in /sys/hypervisor/type
16:47:14 systemd[1]: Virtualization XEN not found, /proc/xen/capabilities does not exist
16:47:14 systemd[1]: Found VM virtualization xen
16:47:14 systemd[1]: Detected virtualization xen.
16:47:15 systemd[1]: haveged.service: ConditionVirtualization=!container succeeded.
16:47:24 systemd[1]: Started Virtualization daemon.
16:47:33 systemd-hostnamed[2202]: No virtualization found in CPUID
16:47:33 systemd-hostnamed[2202]: No virtualization found in DMI
16:47:33 systemd-hostnamed[2202]: Virtualization XEN found (/proc/xen/capabilities exists)
16:47:33 systemd-hostnamed[2202]: Virtualization XEN Dom0 ignored (/proc/xen/capabilities)
16:47:33 systemd-hostnamed[2202]: Found VM virtualization none

Because of the too-early virt-detection and caching of the result, services that expect ConditionVirtualization=false started by systemd[1] fail to start e.g. smartd.service of smartmontools

Other processes run later e.g. systemd-detect-virt also detect correct virt-type for dom0 ("none") as expected.

- can i introduce a dependency on "proc-xen.mount" at systemd startup, in order to avoid this?

- If not, can detection and caching be deferred for later, if VIRTUALIZATION_XEN was detected but /proc/xen/capabilities is not present at all? e.g. example hack:

```
diff --git a/src/basic/virt.c b/src/basic/virt.c
index 601174452..28de9fcb8 100644
--- a/src/basic/virt.c
+++ b/src/basic/virt.c
@@ -253,6 +253,16 @@ static bool detect_vm_xen_dom0(void) {
         return true;
 }

+static bool detect_vm_xen_capabilities_present(void) {
+        if (access("/proc/xen/capabilities", F_OK) < 0) {
+                log_debug("Virtualization XEN procfs not found, /proc/xen/capabilities does not exist");
+                return false;
+        }
+
+        log_debug("Virtualization XEN procfs found (/proc/xen/capabilities exists)");
+        return true;
+}
+
 static int detect_vm_hypervisor(void) {
         _cleanup_free_ char *hvtype = NULL;
         int r;
@@ -381,6 +391,8 @@ int detect_vm(void) {
                 return r;

 finish:
+         /* If xenfs has not been mounted yet, Xen dom0 virtualization detection can be missed.
+          * Set it to invalid so that it is checked again later */
+        if (r == VIRTUALIZATION_XEN && !detect_vm_xen_capabilities_present())
+                r = _VIRTUALIZATION_INVALID;
         /* x86 xen Dom0 is detected as XEN in hypervisor and maybe others.
          * In order to detect the Dom0 as not virtualization we need to
          * double-check it */
```

### In case of bug report: Steps to reproduce the problem

Simply boot SLES12SP2 Xen dom0

**Comment by keszybz** [comment]
> can i introduce a dependency on "proc-xen.mount" at systemd startup, in order to avoid this?

No, because this function is called when systemd is starting up, before any units are mounted.

> If not, can detection and caching be deferred for later, if VIRTUALIZATION_XEN was detected but 

That'd be possible, but ugly and hard to get right.

I'd rather try to get the check to work even in early boot. The easiest option would be to make `/proc/xen` an API filesystem that would be mounted by systemd like /proc and /sys and others. Is there any reason to not do that?

**Comment by poettering** [comment]
Yeah, I think the best would be to ad the xen stuff to mount-setup.c and thus mount it in earliest boot, to fix this issue. Given that xen is open source and the kernel support all merged this should be OK to do.

The patch adding that should be trivial. But I figure we need a patch for that, as we can't test this here, as we don't have Xen systems around. Any chance you can prep something?

The patch should add two lines to mount_table[] in src/core/mount-setup.c and should be trivial.

**Comment by fsateler** [comment]
Shouldn't this be closed now?

**Comment by keszybz** [comment]
Fixed by b305bd3aab0446095bf28ffc28b3d908a7ace6b9.

**Comment by vliaskov** [comment]
Please re-open. This change makes proc-xen.mount obsolete and would need to be removed as a dependency (in upstream xen package). Otherwise this breaks xen. I suggest reverting this change until we know upstream xen would be ok with removing proc-xen.mount.
